backend/server.py

```diff
@@ -211,18 +281,47 @@ def read_segments(source_id: int, user_info: dict = Depends(get_user_info)):
 async def save_segments(request: Request, user_info: dict = Depends(get_user_info)):
     try:
         data = await request.json()
-        segments = data.get("segments", [])
-
-        if not isinstance(segments, list):
-            raise HTTPException(status_code=400, detail="Invalid request format - segments must be a list")
-
-        # Add username and timestamp to each segment
         now = datetime.utcnow()
-        for segment in segments:
-            segment["username"] = user_info["preferred_username"]
-            segment["timestamp"] = now
-
+        username = user_info["preferred_username"]
+        
+        # New format: build segments from paragraphs
+        if "paragraphs" in data:
+            paragraphs = data.get("paragraphs", [])
+            source_id = data.get("source_id")
+            properties = data.get("properties", {})
+            original_segments = data.get("originalSegments", [])
+            additional_sources_segments = data.get("additional_sources_segments", {})
+            
+            segments = build_segments_from_paragraphs(
+                paragraphs, source_id, properties, username, now, original_segments
+            )
+            
+            if additional_sources_segments:
+                segments.extend(build_additional_sources_segments(
+                    additional_sources_segments, properties, username, now
+                ))
+                # Update temporal segments (order=0) by removing consumed text
+                update_temporal_segments_remaining_text(
+                    additional_sources_segments, username, now
+                )
+        else:
+            # Old format: segments already built
```

Comment:
If we have to "switch" between two formats, this means there is a flaw in the design or understanding.